pennylane/templates/subroutines/trotter.py

```diff
@@ -0,0 +1,216 @@
+        evolution time: 
+
+        .. code-block:: python3
+
+            @qml.qnode(dev)
+            def my_circ(c1, c2, time): 
+                # Prepare H: 
+                H = qml.dot([c1, c2], [qml.PauliX(0), qml.PauliZ(1)])
+                    
+                # Prepare some state
+                qml.Hadamard(0)
+                
+                # Evolve according to H
+                qml.TrotterProduct(H, time, order=2)
+                
+                # Measure some quantity
+                return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))
+
+        >>> args = qnp.array([1.23, 4.5, 0.1])
+        >>> qml.grad(my_circ)(*tuple(args))
+        (tensor(0.00961064, requires_grad=True), tensor(-0.12338274, requires_grad=True), tensor(-5.43401259, requires_grad=True))
+    """
+
+    def __init__(self, hamiltonian, time, n=1, order=1, check_hermitian=True, id=None):
+        r"""Initialize the TrotterProduct class"""
+
+        if order <= 0 or order != 1 and order % 2 != 0:
+            raise ValueError(
+                f"The order of a TrotterProduct must be 1 or a positive even integer, got {order}."
+            )
+
+        if isinstance(hamiltonian, qml.Hamiltonian):
+            coeffs, ops = hamiltonian.terms()
+            hamiltonian = qml.dot(coeffs, ops)
+
+        if not isinstance(hamiltonian, Sum):
+            raise TypeError(
+                f"The given operator must be a PennyLane ~.Hamiltonian or ~.Sum got {hamiltonian}"
+            )
+
+        if check_hermitian:
+            for op in hamiltonian.operands:
+                if not op.is_hermitian:
+                    raise ValueError(
+                        "One or more of the terms in the Hamiltonian may not be hermitian"
+                    )
+
+        self._hyperparameters = {"num_steps": n, "order": order, "base": hamiltonian}
+        wires = hamiltonian.wires
+        super().__init__(time, wires=wires, id=id)
+
+    @staticmethod
+    def compute_decomposition(*args, **kwargs):
+        r"""Representation of the operator as a product of other operators (static method).
+
+        .. math:: O = O_1 O_2 \dots O_n.
+
+        .. note::
+
+            Operations making up the decomposition should be queued within the
+            ``compute_decomposition`` method.
+
+        .. seealso:: :meth:`~.Operator.decomposition`.
+
+        Args:
+            *params (list): trainable parameters of the operator, as stored in the ``parameters`` attribute
+            wires (Iterable[Any], Wires): wires that the operator acts on
+            **hyperparams (dict): non-trainable hyperparameters of the operator, as stored in the ``hyperparameters`` attribute
+
+        Returns:
+            list[Operator]: decomposition of the operator
+        """
+        time = args[0]
+        n = kwargs["num_steps"]
+        order = kwargs["order"]
+        ops = kwargs["base"].operands
+
+        decomp = _recursive_expression(time / n, order, ops)[-1::-1] * n
```

Comment:
Very elegant 😄
Is it necessary to go through it from the end?

Reply:
This is a good point, we could potentially generate the list already inverted in circuit order instead of math order. It won't look as elegant, but should be more efficient. Would you be free to help me derive this tomorrow?